block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -133,57 +135,95 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
   def insert(
       block: BlockMessage,
       invalid: Boolean
-  ): F[BlockDagRepresentation[F]] =
-    lock.withPermit(
+  ): F[BlockDagRepresentation[F]] = {
+    import cats.instances.list._
+    import cats.instances.option._
+    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
+
+    // Empty sender is valid for genesis
+    val senderIsEmpty          = block.sender == ByteString.EMPTY
+    val senderHasInvalidFormat = !senderIsEmpty && (block.sender.size() != Validator.Length)
+    val sendersNewLM           = (block.sender, block.blockHash)
+
+    val logAlreadyStored =
+      Log[F].warn(s"Block ${PrettyPrinter.buildString(block, short = true)} is already stored.")
+
+    val logEmptySender =
+      Log[F].warn(s"Block ${Base16.encode(block.blockHash.toByteArray)} sender is empty")
+
+    // Add LM either if there is no existing message for the sender, or if sequence number advances
+    // - assumes block sender is not valid hash
+    def shouldAddAsLatest: F[Boolean] =
```

Comment:
We don't need to check if sender is empty here because this function should be called only for valid sender.